### PR TITLE
CA-330162 Ensure floppy options are passed to qemu

### PR DESCRIPTION
**Pull request by lippirk**
Previously a floppy VBD would be treated as a
disk, which caused xenopsd to filter out that vbd
when deciding what args to execute qemu with.

Signed-off-by: lippirk <ben.anson@citrix.com>

### CODE REVIEW

File: xc/xenops_server_xen.ml [hunk 1]
```diff
@@ -1460,6 +1460,7 @@ module VM = struct
               match vbd.Vbd.ty, vbd.mode with
               | Vbd.Disk, ReadOnly -> None
               | Vbd.Disk, _        -> Some (index, path, Device.Dm.Disk)
+              | Vbd.Floppy, _      -> Some (index, path, Device.Dm.Floppy)
               | _                  -> Some (index, path, Device.Dm.Cdrom)
```
Comment (edwintorok) [inline xc/xenops_server_xen.ml]: I'd replace the wildcard with Vbd.Cdrom to be explicit and avoid treating a floppy as a cdrom by accident.

File: xc/xenops_server_xen.ml [hunk 2]
```diff
@@ -2622,14 +2623,22 @@ module VBD = struct
              let open Device_common in
              let device_number = dev.frontend.devid |> Device_number.of_xenstore_key in
              let qemu_domid = this_domid ~xs in
-             let qemu_frontend = match Device_number.spec device_number with
-               | Ide, n, _ when n < 4 ->
+             let qemu_frontend =
+               let maybe_create_vbd_backend () =
```
Comment (edwintorok) [inline xc/xenops_server_xen.ml]: create vbd frontend for backend? Looking at the code below it creates a frontend, not a backend, based on the presence of a backend.

File: xc/xenops_server_xen.ml [hunk 3]
```diff
@@ -2622,14 +2623,22 @@ module VBD = struct
...
                  end
-               | _,_,_ -> None in
+               in
+               match Device_number.spec device_number with
+               | Ide,    n, _ when 0 <= n && n < 4 -> maybe_create_vbd_backend ()
+               | Floppy, n, _ when 0 <= n && n < 2 -> maybe_create_vbd_backend ()
+               | Ide,    n, _                      -> D.warn "qemu_frontend: Ide index should be less than 4, but got: %i" n; None
+               | Floppy, n, _                      -> D.warn "qemu_frontend: floppy index should be less than 2, but got: %i" n; None
+               | Xen,    _, _                      -> None
+               | Scsi,   _, _                      -> None
```
Comment (edwintorok) [inline xc/xenops_server_xen.ml]: These 2 could be on the same line `(Xen|Scsi)`

File: xc/device.ml [hunk 4]
```diff
@@ -3001,9 +3007,14 @@ module Backend = struct
...
-          info.Dm_Common.disks
-          |> List.partition (function (_, _, Dm_Common.Cdrom) -> true | (_, _, _) -> false)
+        let disks_cdrom, disks_floppy, disks_other =
+          let acc (cdroms, floppies, other) ((_, _, media) as next) =
+            match media with
+            | Dm_Common.Cdrom  -> (next :: cdroms, floppies, other)
+            | Dm_Common.Floppy -> (cdroms, next :: floppies, other)
+            | Dm_Common.Disk   -> (cdroms, floppies, next :: other)
+          in
+          List.fold_left acc ([], [], []) info.Dm_Common.disks
```
Comment (psafont) [inline xc/device.ml]: The fold reverses the order of the elements in the resulting lists. Not sure if it's relevant.

I recommend to change the name from `acc` to something like `partition`.

File: xc/xenops_server_xen.ml (hunk 2 above)
Comment (lippirk) [reply xc/xenops_server_xen.ml]: oops

File: xc/device.ml (hunk 4 above)
Comment (lippirk) [reply xc/device.ml]: hopefully not, all this function does is create a list of args for qemu

partition is indeed a sensible name

File: xc/xenops_server_xen.ml [hunk 5]
```diff
@@ -2622,14 +2623,22 @@ module VBD = struct
...
                    | None   -> Some (index, Empty)
                    | Some _ -> Some (index, create_vbd_frontend ~xc ~xs task qemu_domid vdi)
                  end
-               | _,_,_ -> None in
+               in
+               match Device_number.spec device_number with
+               | Ide,    n, _ when 0 <= n && n < 4 -> maybe_create_vbd_backend ()
+               | Floppy, n, _ when 0 <= n && n < 2 -> maybe_create_vbd_backend ()
+               | Ide,    n, _                      -> D.warn "qemu_frontend: Ide index should be less than 4, but got: %i" n; None
+               | Floppy, n, _                      -> D.warn "qemu_frontend: floppy index should be less than 2, but got: %i" n; None
```
Comment (psafont) [inline xc/xenops_server_xen.ml]: Consider changing message to something similar to "qemu_frontend: XXXX supports device numbers between 0 and Y; received %i.". This message shows both limits, uses terminology from the domain and avoid using "should" when we mean it's not supported at all.

File: xc/device.ml [hunk 6]
```diff
@@ -1863,8 +1863,12 @@ module Dm_Common = struct
     | VNC of disp_intf_opt * string option * bool * int * string option (* IP address, auto-allocate, port if previous false, keymap *)
     | SDL of disp_intf_opt * string (* X11 display *)
 
-  type media = Disk | Cdrom
-  let string_of_media = function Disk -> "disk" | Cdrom -> "cdrom"
+  type media = Disk | Cdrom | Floppy
```
Comment (lindig) [inline xc/device.ml]: I understand that this is an incremental change but I would have liked to move this into a module to regain some structure:

```
module Media = struct
  type t = Disk | Cdrom | Floppy

  let to_string = function
    | Disk -> "disk" ..
   ..

  ..
end
```

File: xc/device.ml [hunk 7]
```diff
@@ -3010,9 +3014,9 @@ module Backend = struct
         let disks_cdrom, disks_floppy, disks_other =
           let partition (cdroms, floppies, other) ((_, _, media) as next) =
             match media with
-            | Dm_Common.Cdrom  -> (next :: cdroms, floppies, other)
-            | Dm_Common.Floppy -> (cdroms, next :: floppies, other)
-            | Dm_Common.Disk   -> (cdroms, floppies, next :: other)
+            | Dm_Common.Media.Cdrom  -> (next :: cdroms, floppies, other)
```
Comment (psafont) [inline xc/device.ml]: You can add `let open Dm_Common.Media` just before the `match` if you want to keep the indentation level of the arrows

File: xc/xenops_server_xen.ml [hunk 8]
```diff
@@ -269,7 +269,7 @@ let params_of_backend backend =
   in
   (params, xenstore_data, extra_keys)
 
-let create_vbd_frontend ~xc ~xs task frontend_domid vdi =
+let create_vb_backend ~xc ~xs task frontend_domid vdi =
```
Comment (lippirk) [inline xc/xenops_server_xen.ml]: got a bit carried away renaming here

**Pull request merged**